### ISSUE TITLE
Tabbing in chat shows BTLP_Slot_1, 2 etc.

### ISSUE THREAD

**Issue by DanielAlves123** [CodeCrafter47/BungeeTabListPlus#384]
Put space in chat and press tab https://cl.ly/b72f1e65f7c8

**Comment by CodeCrafter47** [comment]
I've had a look at what causes this and I don't think I can fix it.

**Comment by mibby** [comment]
@CodeCrafter47 Any possible update on this to disable tab complete for BTLP slots? :(

**Comment by CodeCrafter47** [comment]
I've now implemented my first suggestion, i.e. when pressing tab in chat real players are now displayed above the BTLP slots.

Regarding my other ideas:

- Replacing the slot names with smileys is probably a good way to turn this bug into a useful feature, however it would not be obvious to users which plugin causes this so I'm not sure if I want to do this. Let me know what you think.

- With command tab completion the only command I found where the btlp slots showed up is `/team`. So it's probably not a big issue.

**Comment by mibby** [comment]
@CodeCrafter47 BTLP dev 810 causes quite a few issues. Crashing as soon as I login.

https://paste.ubuntu.com/p/wPX2CsStCZ/

Downgrading to BTLP dev 808, all is fine. https://github.com/CodeCrafter47/BungeeTabListPlus/commit/43bf88b856c9c54e02aa246795c234e184f17e71 a broken commit? :(

**Comment by CodeCrafter47** [comment]
Should be fixed now.

**Comment by mibby** [comment]
Crash confirmed fixed. 

Tab also seems to sort me above all the ~BTLP Slot ## now as well. It's a shame those fake player slots can't be removed from tab altogether, but sorting players above them with them having a tilde in their name seems acceptable. Much better than before!

**Comment by AeSix** [comment]
So, this now occurs without any / or space in the chat, it appears as soon as <tab> is pressed.

Could these fake players be hidden like real players with a Vanish plugin?  It's freaking out my players lol.   (I tried to /vanish one of them, and it didn't work)

Could pre-pending the names with a : solve the issue?  Hacky, but might work as Bungee don't send command (I know it's not a command) to tab completions... worth a shot maybe?

**Comment by CodeCrafter47** [comment]
These cannot be hidden. I do not think using a `:` would fix the issue. How would that be different from the `~` that is currently used?

What do you think about my suggestion of replacing the slot names with smileys? That is definitely possible and will turn something annoying into a (potentially) useful feature.

**Comment by AeSix** [comment]
The only problem I foresee with using smilies and other ascii art, is that you'll need to figure out how many you'll need... I think you'd need a lot more than 60?  I'm not sure what a sane amount of slots to show is.  

Other than that, I like it.  

As far as the :, I was thinking maybe bungee might see it and not send it, but I have no confirmation of this, or any way to test, so, a grasping at straws suggestion. 

I'm wondering though, could ProtocolLib be used, as an optional, to re-build the player list, where you would exclude players starting with "~BTLP"?  I'm not certain that'd work, but it might.  I'll see if my wife, who is actually a programmer, has any ideas when she gets off work.

**Comment by AeSix** [comment]
by the way, this is probably obvious to the old timers and devs, but changing from "FIXED_SIZE" to "DYNAMIC_SIZE" removes the slot players, and so it's not an issue... but at the cost of losing the coolies.  

Posting more for anyone else who happens upon the ticket :)

**Comment by toxamin** [comment]
doesn't seem too difficult https://www.youtube.com/watch?v=7EqMMcyRHsM

**Comment by AeSix** [comment]
That's using BTLP with "DYNAMIC_SIZE"?
How'd you disappear the fake players?  There's a PR? :D
Would be wonderful to have that 💯

**Comment by toxamin** [comment]
No, it's not BTLP. It's a PoC, simply proof it's possible. I thought this could be applied to BTLP, however, I'm not sure if major parts of the code base have to be rewritten for this to work.

**Comment by AeSix** [comment]
Ahh, makes sense!  Thank you for your efforts! :)

**Comment by CodeCrafter47** [comment]
@toxamin How did you prevent the fake players from showing up in tab completion? Is it something specific to 1.15 (haven't looked at that yet) or would it also work for older versions?

**Comment by DanielAlves123** [comment]
@toxamin @CodeCrafter47 I would love to see this fixed! <3

**Comment by toxamin** [comment]
To put it simply, you can't tab complete nothing (aka "").
You can add multiple entries to the tablist with the name "", because in 1.8+, they are identified by their uuids, and not their names.
If there are multiple entries with the same name, they seem to be sorted by uuid.
So, the solution is to add players with the name "", with unique uuids.
This seems to work in 1.8+ versions, however, I have only tested simple things.
The blank entries are sorted above normal players (which is good, because I'm not sure if a blank entry can be added to a team).

**Comment by CodeCrafter47** [comment]
That seems like a good option. It will only work when using all 80 slots, but that is fine. I will look into implementing this next year.

**Comment by mibby** [comment]
@CodeCrafter47 Any possible update on looking into this? :)

**Comment by CodeCrafter47** [comment]
I am currently working on a big update fixing most issues on the [Far Future ](https://github.com/CodeCrafter47/BungeeTabListPlus/milestone/6)milestone. It is mostly done, except for a bit of testing/ fixing some bugs. I will look into this once that is done.

**Comment by mibby** [comment]
Great, hopefully it's an easy fix! For reference, this is what shows up now with dev build 848. ![link](https://i.imgur.com/KHe5Uq1.png)

**Comment by mibby** [comment]
Is this something that can be revisited in the future? Fake players still show up in the tab complete at the very bottom with BTLP dev 985, and with Geyser (Bedrock Client Support), players see this.

![link-br](https://cdn.discordapp.com/attachments/516337315900030977/701624702334009434/Screenshot_20200419-224530.jpg)

**Comment by CodeCrafter47** [comment]
I definitely want to give this another try in the future.

I am not particularly worried about the bedrock support though.

**Comment by mibby** [comment]
No worries, I'll be here to help test when you experiment with future changes to try and nail down this problem. :) 

Bedrock support would technically be fixed and go hand-in-hand with the fake players no longer being included in the tab-complete list, but obviously it's nothing to worry about since it's an unofficial compatibility hack anyways.

**Comment by EnderGamingFilms** [comment]
So is this issue patched when player size is 80 or was this patch abandoned for the time being?

**Comment by CodeCrafter47** [comment]
The patch has been reverted. I plan on adding it back for the next update, with a config option associated to it, so people can enable/ disable it depending on how well it works for them.

**Comment by CodeCrafter47** [comment]
I've added the `experimentalTabCompleteFixForTabSize80` and `experimentalTabCompleteSmileys` options to the `config.yml`. The first one removes all the `~BTLP` entries from tab completion if the size of the tab list is 80. The second option can be used to replace the `~BTLP` entries with some semi-useful unicode characters. You may use both options at the same time. Let me know how it goes.

**Comment by mibby** [comment]
@CodeCrafter47 Setting ``experimentalTabCompleteFixForTabSize80: true`` breaks the tab list for me. 

![link](https://i.imgur.com/MC95mmz.png)

Had it happen when I switched servers. Also had it happen to someone who logged in / connected. Relogging usually fixes the issue. However I'm able to force it to break by running ``/btlp reload`` from Bungee.

Gallery of different breakage.
https://imgur.com/a/2G0FFev

How it looks normally.
https://i.imgur.com/YSrEAL7.png

Turning the setting off (false) fixes the problem.

BungeeTabListPlus dev 1007

**Comment by CodeCrafter47** [comment]
Doesn't happen on my test server. I suspect you might need some other plugins to make the issue happen. Maybe you can try to narrow it down a bit.

**Comment by mibby** [comment]
The only other plugin I have that modifies scoreboards is ``TAB``, which I only use to set player nametag colors (name above head). It does not auto refresh and all other modules besides nametags are disabled. It is also server specific and not on the proxy like BTLP, so I'm not too sure why BTLP would break when reloading BTLP from Bungee while standing still on the server, with the option enabled.

https://www.spigotmc.org/resources/tab-1-5-x-1-15-x-free-version.57806/
https://github.com/NEZNAMY/TAB

I'll see if it breaks on my test server part of the same proxy network with minimal plugins.

**Comment by mibby** [comment]
@CodeCrafter47 Apologies for getting back to you so late! Upon re-testing BTLP on my same network but on my test server with minimal plugins and no plugins that modify scoreboard groups or nametags, BTLP still breaks when ``experimentalTabCompleteFixForTabSize80: true`` is set.

BTLP dev 1016
Waterfall dev 346 (BungeeCord Fork)
Paper dev 348 (Spigot 1.15.2)

![link50](https://i.imgur.com/KKNcMOf.png)

**Comment by CodeCrafter47** [comment]
Unfortunately I am still unable to reproduce the issue, even using these exact versions. However with the next update the experimental options will be available to a wider audience, maybe that will help in tracking down the issue.

**Comment by EnderGamingFilms** [comment]
How do I use this feature I don't see the options in config.yml

**Comment by loperd** [comment]
And when will the next update be available?

**Comment by CodeCrafter47** [comment]
I intend to publish the next update shortly after the Minecraft 1.16 release.

**Comment by CodeCrafter47** [comment]
If anyone could have a look at [the latest dev-build ](https://ci.codecrafter47.de/job/BungeeTabListPlus/)to test the commit I made two days ago that would be nice. It should fix the issue of players showing up at the top left of the tab list when the experimentalTabCompleteFixForTabSize80 option is enabled. At least it fixes an instance of the issue which I managed to reproduce. I my case I had LibsDisguises on the spigot server which triggered the issue. The fix is rather generic, so it should also work if a different plugin triggers the issue in a similar way.

**Comment by Prouser123** [comment]
@CodeCrafter47 We aren't using LibsDisguises but the dev build works perfectly for us!

**Comment by AlexanderMoebius** [comment]
What about changing (or faking) some of these ids to the names of the players that are online at another server?

That would also enable server wide chat tab complete with player names.

**Comment by Toseflo** [comment]
That's also what I thought. Would be really nice to have server-wide tab complete!

**Comment by Andre601** [comment]
Something I would like to know is, how BTLP actually handles this hiding of fake players on size 80 tab lists.

Wouldn't it be a possible solution to check for certain packets used in tab completion and manipulate those, so that the fake players are always removed?
Like, couldn't such behaviour even work on smaller tab lists? I'm pretty sure it's doable to listen for those packets and then manipulate them in the right way so that the player only sees the real players in the completion (And considering that this only happens in tab and the player, therefore, can't see the tab would I not even worry that much about the tab looking messed up or something)
